test/functional/webauthn_verifications_controller_test.rb

```diff
@@ -69,4 +68,118 @@ class WebauthnVerificationsControllerTest < ActionController::TestCase
       end
     end
   end
+
+  context "#authenticate" do
+    setup do
+      @user = create(:user)
+      @webauthn_credential = create(:webauthn_credential, user: @user)
+      @token = create(:webauthn_verification, user: @user).path_token
+      get :prompt, params: { webauthn_token: @token }
+    end
+
+    context "when verifying the challenge" do
+      setup do
+        @challenge = session[:webauthn_authentication]["challenge"]
+        @origin = "http://localhost:3000"
+        @rp_id = URI.parse(@origin).host
+        @client = WebAuthn::FakeClient.new(@origin, encoding: false)
+        WebauthnHelpers.create_credential(
+          webauthn_credential: @webauthn_credential,
+          client: @client
+        )
+        post(
+          :authenticate,
+          params: {
+            credentials:
+              WebauthnHelpers.get_result(
+                client: @client,
+                challenge: @challenge
+              ),
+            webauthn_token: @token
+          },
+          format: :json
+        )
+      end
+
+      should respond_with :success
+      should "return success message" do
+        assert_equal "success", JSON.parse(response.body)["message"]
+      end
+    end
+
+    context "when not providing credentials" do
+      setup do
+        post(
+          :authenticate,
+          params: {
+            webauthn_token: @token
+          },
+          format: :json
+        )
+      end
+
+      should respond_with :unauthorized
+      should "return error message" do
+        assert_equal "Credentials required", JSON.parse(response.body)["message"]
+      end
+    end
+
+    context "when providing wrong credentials" do
+      setup do
+        @wrong_challenge = "16b8e11ea1b46abc64aea3ecdac1c418"
+        @origin = "http://localhost:3000"
+        @rp_id = URI.parse(@origin).host
+        @client = WebAuthn::FakeClient.new(@origin, encoding: false)
+        WebauthnHelpers.create_credential(
+          webauthn_credential: @webauthn_credential,
+          client: @client
+        )
+        post(
+          :authenticate,
+          params: {
+            credentials:
+              WebauthnHelpers.get_result(
+                client: @client,
+                challenge: @wrong_challenge
+              ),
+            webauthn_token: @token
+          },
+          format: :json
+        )
+      end
+
+      should respond_with :unauthorized
+      should "return error message" do
+        assert_equal "WebAuthn::ChallengeVerificationError", JSON.parse(response.body)["message"]
```

Comment:
`WebAuthn::ChallengeVerificationError` isn't the best error message, I don't think it would be triggered much but could be worth punting it as an future enhancement